tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestFunBuilder.scala

```diff
@@ -0,0 +1,288 @@
+package at.forsyte.apalache.tla.typecomp
+
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.lir.oper.TlaFunOper
+import org.junit.runner.RunWith
+import org.scalacheck.Gen
+import org.scalatestplus.junit.JUnitRunner
+import scalaz.unused
+
+@RunWith(classOf[JUnitRunner])
+class TestFunBuilder extends BuilderTest {
+
+  test("tuple") {
+    type T = Seq[TBuilderInstruction]
+
+    type TParam = Seq[TlaType1]
+
+    implicit val typeSeqGen: Gen[TParam] = for {
+      n <- Gen.choose(0, 5)
+      seq <- Gen.listOfN(n, singleTypeGen)
+    } yield seq
+
+    def mkWellTyped(tparam: TParam): T =
+      tparam.zipWithIndex.map { case (tt, i) =>
+        builder.name(s"t$i", tt)
+      }
+
+    def mkIllTyped(@unused tparam: TParam): Seq[T] = Seq.empty
+
+    def resultIsExpected = expectEqTyped[TParam, T](
+        TlaFunOper.tuple,
+        mkWellTyped,
+        liftBuildToSeq,
+        ts => TupT1(ts: _*),
+    )
+
+    checkRun(
+        runVariadic(
+            builder.tuple,
+            mkWellTyped,
+            mkIllTyped,
+            resultIsExpected,
+        )
+    )
+
+  }
+
+  test("funDef") {
+    type T = (TBuilderInstruction, TBuilderInstruction, TBuilderInstruction)
+
+    type TParam = (TlaType1, TlaType1)
+
+    def mkWellTyped(tparam: TParam): T = {
+      val (t, tt) = tparam
+      (
+          builder.name("e", t),
+          builder.name("x", tt),
+          builder.name("S", SetT1(tt)),
+      )
+    }
+
+    def mkIllTyped(tparam: TParam): Seq[T] = {
+      val (t, tt) = tparam
+      Seq(
+          (
+              builder.name("e", t),
+              builder.name("x", InvalidTypeMethods.differentFrom(tt)),
+              builder.name("S", SetT1(tt)),
+          ),
+          (
+              builder.name("e", t),
+              builder.name("x", tt),
+              builder.name("S", InvalidTypeMethods.notSet),
+          ),
+          (
+              builder.name("e", t),
+              builder.name("x", tt),
+              builder.name("S", SetT1(InvalidTypeMethods.differentFrom(tt))),
+          ),
+      )
+    }
+
+    def resultIsExpected = expectEqTyped[TParam, T](
+        TlaFunOper.funDef,
+        mkWellTyped,
+        { case (a, b, c) => Seq(a, b, c) },
+        { case (a, b) => FunT1(b, a) },
+    )
+
+    checkRun(
+        runTernary(
+            builder.funDef,
+            mkWellTyped,
+            mkIllTyped,
+            resultIsExpected,
+        )
+    )
```

Comment:
I am not sure I understand what is going on here. Please add some comments, since this testing framework is non-obvious.